sklearn/neighbors/binary_tree.pxi

```diff
@@ -1164,7 +1164,11 @@ cdef class BinaryTree:
         self.dist_metric = state[11]
         self.euclidean = (self.dist_metric.__class__.__name__
                           == 'EuclideanDistance')
-        self.sample_weight = state[12]
+        self.sample_weight_arr = state[12]
+
+        if self.sample_weight_arr is not None:
+            self.sample_weight = get_memview_DTYPE_1D(self.sample_weight_arr)
+            self.sum_weight = np.sum(self.sample_weight)
 
```

Comment:
You need to, like `__init__` set `sum_weight` to the number of samples when sample_weight is None. Perhaps refactor this reinitialisation code that is duplicated in init

Reply:
@jnothman Actually you are correct. Now just the test_pickling is failing and not the new test. Thanks for the comment! :)